sway/server.c

```diff
@@ -101,3 +124,289 @@ void server_run(struct sway_server *server) {
+		wlr_headless_backend_create(server->wl_display);
+	return backend;
+}
+
+static struct wlr_backend *drm_backend_create(struct sway_server *server) {
+	wlr_log(L_DEBUG, "TODO: create drm backend");
+	return NULL;
+}
+
+void sway_server_add_subbackend(struct sway_server *server,
+		struct sway_subbackend *subbackend) {
+	if (sway_server_get_subbackend(server, subbackend->name)) {
+		wlr_log(L_ERROR, "cannot add subbackend '%s': already exists",
+			subbackend->name);
+		sway_subbackend_destroy(subbackend);
+		return;
+	}
+
+	struct wlr_backend *backend = NULL;
+
+	switch (subbackend->type) {
+	case SWAY_SUBBACKEND_WAYLAND:
+		backend = wayland_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_X11:
+		backend = x11_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_DRM:
+		backend = drm_backend_create(server);
+		break;
+	case SWAY_SUBBACKEND_HEADLESS:
+		backend = headless_backend_create(server);
+		break;
+	}
+
+	if (backend == NULL) {
+		wlr_log(L_ERROR, "could not create subbackend '%s'", subbackend->name);
+		sway_subbackend_destroy(subbackend);
+		return;
+	}
+
+	subbackend->backend = backend;
+
+	wl_list_remove(&subbackend->link);
+	wl_list_insert(&server->subbackends, &subbackend->link);
+
+	wl_signal_add(&backend->events.destroy, &subbackend->backend_destroy);
+	subbackend->backend_destroy.notify = handle_subbackend_backend_destroy;
+
+	wlr_multi_backend_add(server->backend, backend);
+	wlr_backend_start(backend);
+}
+
+void sway_server_remove_subbackend(struct sway_server *server, char *name) {
+	struct sway_subbackend *subbackend =
+		sway_server_get_subbackend(server, name);
+
+	if (!subbackend) {
+		wlr_log(L_DEBUG, "could not find subbackend named '%s'", name);
+		return;
+	}
+
+	wlr_backend_destroy(subbackend->backend);
+}
+
+struct subbackend_output {
+	struct sway_subbackend *backend;
+	struct wlr_output *wlr_output;
+
+	struct wl_listener output_destroy;
+
+	struct wl_list link; // sway_subbackend::outputs
+};
+
+static void subbackend_output_destroy(struct subbackend_output *output) {
+	wl_list_remove(&output->link);
+	wl_list_remove(&output->output_destroy.link);
+	free(output);
+}
+
+static void handle_subbackend_output_destroy(struct wl_listener *listener,
+		void *data) {
+	struct subbackend_output *output =
+		wl_container_of(listener, output, output_destroy);
+	subbackend_output_destroy(output);
+}
+
+void sway_subbackend_add_output(struct sway_server *server,
+		struct sway_subbackend *subbackend, char *name) {
+	struct wlr_output *wlr_output = NULL;
+
+	switch(subbackend->type) {
+	case SWAY_SUBBACKEND_WAYLAND:
+		wlr_log(L_DEBUG, "TODO: create wayland subbackend output");
+		break;
+	case SWAY_SUBBACKEND_X11:
+		wlr_log(L_DEBUG, "TODO: create x11 subbackend output");
+		break;
+	case SWAY_SUBBACKEND_DRM:
+		wlr_log(L_DEBUG, "creating DRM subbackend outputs is not supported");
+		break;
+	case SWAY_SUBBACKEND_HEADLESS:
+		wlr_output =
+			wlr_headless_add_output(subbackend->backend, 1280, 960);
```

Comment:
I know I'm picking nits here but we should probably default to 1280x720. Also, we should have `add-output headless-2 [WIDTHxHEIGHT|WIDTH HEIGHT]`

Reply:
I wonder if we can set the mode on headless outputs to an arbitrary resolution via `output HEADLESS-2 ...`, hmm...